File: daily challenge/MergeSort.py
class Solution(object):
	def sorted(self, A):
		tmp = [0 for _ in range(len(A))]
		start, end = 0, len(A) - 1
		self.mergesort(A, start, end)

	def mergesort(self, A, start, end):
		if start >= end:
			return
		mid = (start + end) // 2
		self.mergesort(A,start,mid,tmp)
		self.mergesort(A,mid+1,start,tmp)
		self.merge(start,mid,end,A,tmp)

	def merge(self,start,mid,end,A,tmp):
		leftIdx = start
		rightIdx = mid + 1
		index = start
		if leftIdx <= mid and rightIdx <= end:
			if A[leftIdx] < A[rightIdx]:
				tmp[index] = A[leftIdx]
				leftIdx += 1
			else:
				tmp[index] = A[rightIdx]
				rightIdx += 1

			index += 1

		while leftIdx <= mid:
			tmp[index] = A[leftIdx]
			leftIdx+= 1
			index += 1

		while rightIdx <= mid:
			tmp[index] = A[rightIdx]
			rightIdx += 1
			index += 1

		A = tmp.copy()


		

File: C++ Version/MergeSort.cpp
void mergeSort(int* a, int left, int right){
  if (left < right) {
    int mid = left + (right - left) / 2;
    mergeSort(a, left, mid);
    mergeSort(a, mid+1, right);
    merge(a, left, right, mid);
  }
}

void merge(int* a, int left, int right, int mid){
  int sorted[right-left+1];
  int i = left, j = mid+1;
  int idx = 0;
  while (i<=mid && j<=right) {
    if (a[i] < a[j]) {
      sorted[idx] = a[i];
      idx += 1;
      i += 1;
    }
    else{
      sorted[idx] = a[j];
      j += 1;
      idx += 1;
    }
  }

  while (i<= mid) {
    sorted[idx] = a[i];
    idx += 1;
    i += 1;
  }
  while (j<= right) {
    sorted[idx] = a[j];
    idx += 1;
    j += 1;
  }
  for (int i = left, q = 0; i <= right; i++, q++) {
    a[i] = sorted[q];
  }

}
